pmempool: fixes transform with repeated paths

wojtuss opened the pull request:
Ref: pmem/issues#285

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/nvml/1265)

<!-- Reviewable:end -->

krzycz commented:
Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[src/libpmempool/transform.c, line 112 at r1](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTK8Q8p58-I1B6jcfLB:-KTK8Q8p58-I1B6jcfLC:bsuo7xm) ([raw file](https://github.com/pmem/nvml/blob/e266c9e1f718b03047c6ee24e5dab384cb14104e/src/libpmempool/transform.c#L112)):*
> ```C
> check_paths_repeat(struct pool_set *set)
> {
> 	for (unsigned r1 = 0; r1 < set->nreplicas; ++r1) {
> ```

` ... < set->nreplicas - 1` ?  (be careful with nreplicas == 1)

---

*[src/libpmempool/transform.c, line 114 at r1](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTK81vLsTrMft-NAPuk:-KTK81vLsTrMft-NAPul:b-m5s69z) ([raw file](https://github.com/pmem/nvml/blob/e266c9e1f718b03047c6ee24e5dab384cb14104e/src/libpmempool/transform.c#L114)):*
> ```C
> 	for (unsigned r1 = 0; r1 < set->nreplicas; ++r1) {
> 		struct pool_replica *rep1 = set->replica[r1];
> 		for (unsigned r2 = 0; r2 < set->nreplicas; ++r2) {
> ```

`r2 = r1 + 1; ...` ?

---

*[src/libpmempool/transform.c, line 118 at r1](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTK99mk4SZ5vWUcdkvk:-KTK99mk4SZ5vWUcdkvl:bgljgap) ([raw file](https://github.com/pmem/nvml/blob/e266c9e1f718b03047c6ee24e5dab384cb14104e/src/libpmempool/transform.c#L118)):*
> ```C
> 			for (unsigned p1 = 0; p1 < rep1->nparts; ++p1) {
> 				for (unsigned p2 = 0; p2 < rep2->nparts; ++p2) {
> 					if ((r1 != r2 || p1 != p2) &&
> ```

'p1 != p2` ? why?

---

*[src/libpmempool/transform.c, line 120 at r1](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTK9IURXLc82nKIx6yW:-KTK9IURXLc82nKIx6yX:bla3txf) ([raw file](https://github.com/pmem/nvml/blob/e266c9e1f718b03047c6ee24e5dab384cb14104e/src/libpmempool/transform.c#L120)):*
> ```C
> 					if ((r1 != r2 || p1 != p2) &&
> 						strcmp(PART(rep1, p1).path,
> 						PART(rep2, p2).path) == 0)
> ```

something is wrong with indentation here

---

*[src/libpmempool/transform.c, line 156 at r1](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTK9SjcuyhzUHqz2uDb:-KTK9SjcuyhzUHqz2uDc:bxtufgf) ([raw file](https://github.com/pmem/nvml/blob/e266c9e1f718b03047c6ee24e5dab384cb14104e/src/libpmempool/transform.c#L156)):*
> ```C
> 	 * check if part files do not repeat
> 	 */
> 	if (check_paths_repeat(set_out)) {
> ```

couldn't we do it as a part of check_parts_dir()?

---

*[src/test/pmempool_transform/out7.log.match, line 17 at r1](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTK71Vu6_TjyxzYabLJ:-KTK71VyvGZ8PenN9ZzG:bxfpjz0) ([raw file](https://github.com/pmem/nvml/blob/e266c9e1f718b03047c6ee24e5dab384cb14104e/src/test/pmempool_transform/out7.log.match#L17)):*
> ```
> 
> error: Invalid argument
> error: failed to transform /tmp/test_pmempool_transform7/poolset.in -> /tmp/test_pmempool_transform7/poolset.out: Transformation failed
> ```

hardcoded paths - use $(nW) tags here

---

*[src/test/pmempool_transform/out7.log.match, line 32 at r1](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTK7EicAkKOKUgIDkqM:-KTK7EicAkKOKUgIDkqN:b-jzuqa1) ([raw file](https://github.com/pmem/nvml/blob/e266c9e1f718b03047c6ee24e5dab384cb14104e/src/test/pmempool_transform/out7.log.match#L32)):*
> ```
> Heap offset              : 0x302000
> Heap size                : 17817600
> Checksum                 : 0x315a2e0aafe3fbf0 [OK]
> ```

Don't you need $(nW) here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

lplewa commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[src/libpmempool/transform.c, line 114 at r1](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTK9LBTcASplMtSfR3J:-KTK9LBTcASplMtSfR3K:by5jao2) ([raw file](https://github.com/pmem/nvml/blob/e266c9e1f718b03047c6ee24e5dab384cb14104e/src/libpmempool/transform.c#L114)):*
> ```C
> 	for (unsigned r1 = 0; r1 < set->nreplicas; ++r1) {
> 		struct pool_replica *rep1 = set->replica[r1];
> 		for (unsigned r2 = 0; r2 < set->nreplicas; ++r2) {
> ```

something like r2=r1+1 will be more optimal

---

*[src/test/pmempool_transform/out7.log.match, line 17 at r1](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTK7mLJQyy29AVECY83:-KTK7mLKvDi75IeA9lUS:bp5nmda) ([raw file](https://github.com/pmem/nvml/blob/e266c9e1f718b03047c6ee24e5dab384cb14104e/src/test/pmempool_transform/out7.log.match#L17)):*
> ```
> 
> error: Invalid argument
> error: failed to transform /tmp/test_pmempool_transform7/poolset.in -> /tmp/test_pmempool_transform7/poolset.out: Transformation failed
> ```

hardcoded path

---

*[src/test/pmempool_transform/TEST7, line 56 at r1](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTK7rs1UWHPNy2Pnk3A:-KTK7rs1UWHPNy2Pnk3B:bz1aufn) ([raw file](https://github.com/pmem/nvml/blob/e266c9e1f718b03047c6ee24e5dab384cb14104e/src/test/pmempool_transform/TEST7#L56)):*
> ```
> 
> # Create poolset files
> cat << EOF > $POOLSET_IN
> ```

Please use create poolset function..... make windows port easier.... 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

wojtuss commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[src/libpmempool/transform.c, line 112 at r1](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTK8Q8p58-I1B6jcfLB:-KTNaaHroPstqG97MgvE:b-zgh9gy) ([raw file](https://github.com/pmem/nvml/blob/e266c9e1f718b03047c6ee24e5dab384cb14104e/src/libpmempool/transform.c#L112)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> ` ... < set->nreplicas - 1` ?  (be careful with nreplicas == 1)
</details>
each replica has to be compared also against itself

---

*[src/libpmempool/transform.c, line 114 at r1](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTK9LBTcASplMtSfR3J:-KTNWF0ByOcbHnggAnW0:b-cms5dy) ([raw file](https://github.com/pmem/nvml/blob/e266c9e1f718b03047c6ee24e5dab384cb14104e/src/libpmempool/transform.c#L114)):*
<details><summary><i>Previously, lplewa (Łukasz Plewa) wrote…</i></summary>
> something like r2=r1+1 will be more optimal
</details>
yeah, actually r2=r1, because r2 and r1 have to be compared too

---

*[src/libpmempool/transform.c, line 114 at r1](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTK81vLsTrMft-NAPuk:-KTNWPbJP0HJbmb4Yc3X:b-cms5dy) ([raw file](https://github.com/pmem/nvml/blob/e266c9e1f718b03047c6ee24e5dab384cb14104e/src/libpmempool/transform.c#L114)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> `r2 = r1 + 1; ...` ?
</details>
yeah, actually r2=r1, because r2 and r1 have to be compared too

---

*[src/libpmempool/transform.c, line 118 at r1](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTK99mk4SZ5vWUcdkvk:-KTNWTA2p1a-yToqCxKP:b-rb10ox) ([raw file](https://github.com/pmem/nvml/blob/e266c9e1f718b03047c6ee24e5dab384cb14104e/src/libpmempool/transform.c#L118)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> 'p1 != p2` ? why?
</details>
because for r1==r2 and p1==p2 paths will be the same

---

*[src/libpmempool/transform.c, line 120 at r1](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTK9IURXLc82nKIx6yW:-KTNWnhXdBVRJaITtiIK:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/e266c9e1f718b03047c6ee24e5dab384cb14104e/src/libpmempool/transform.c#L120)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> something is wrong with indentation here
</details>
Done.

---

*[src/libpmempool/transform.c, line 156 at r1](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTK9SjcuyhzUHqz2uDb:-KTNXAwL50cPD8BuSjtV:bpt41x0) ([raw file](https://github.com/pmem/nvml/blob/e266c9e1f718b03047c6ee24e5dab384cb14104e/src/libpmempool/transform.c#L156)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> couldn't we do it as a part of check_parts_dir()?
</details>
yes, we can

---

*[src/test/pmempool_transform/out7.log.match, line 17 at r1](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTK71Vu6_TjyxzYabLJ:-KTNamHB7cruHeBJ-A8S:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/e266c9e1f718b03047c6ee24e5dab384cb14104e/src/test/pmempool_transform/out7.log.match#L17)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> hardcoded paths - use $(nW) tags here
</details>
Done.

---

*[src/test/pmempool_transform/out7.log.match, line 17 at r1](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTK7mLJQyy29AVECY83:-KTNan01edBetKKcsUam:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/e266c9e1f718b03047c6ee24e5dab384cb14104e/src/test/pmempool_transform/out7.log.match#L17)):*
<details><summary><i>Previously, lplewa (Łukasz Plewa) wrote…</i></summary>
> hardcoded path
</details>
Done.

---

*[src/test/pmempool_transform/out7.log.match, line 32 at r1](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTK7EicAkKOKUgIDkqM:-KTNaoF8du337SmdFOUP:b32wsm) ([raw file](https://github.com/pmem/nvml/blob/e266c9e1f718b03047c6ee24e5dab384cb14104e/src/test/pmempool_transform/out7.log.match#L32)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Don't you need $(nW) here?
</details>
no

---

*[src/test/pmempool_transform/TEST7, line 56 at r1](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTK7rs1UWHPNy2Pnk3A:-KTNasqoyQwN4xL2NOgE:b-l1kgls) ([raw file](https://github.com/pmem/nvml/blob/e266c9e1f718b03047c6ee24e5dab384cb14104e/src/test/pmempool_transform/TEST7#L56)):*
<details><summary><i>Previously, lplewa (Łukasz Plewa) wrote…</i></summary>
> Please use create poolset function..... make windows port easier.... 
</details>
No, because the create poolset function changes paths of the parts (value of $DIR can differ inside this function).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

lplewa commented:
Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/test/pmempool_transform/TEST7, line 56 at r1](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTK7rs1UWHPNy2Pnk3A:-KTi8mdkSGUsIHlZR5tH:bnvwaxx) ([raw file](https://github.com/pmem/nvml/blob/e266c9e1f718b03047c6ee24e5dab384cb14104e/src/test/pmempool_transform/TEST7#L56)):*
<details><summary><i>Previously, wojtuss (Wojciech Uss) wrote…</i></summary>
> No, because the create poolset function changes paths of the parts (value of $DIR can differ inside this function).
</details>
Add next functionality to create pool set function to don't modify PATH if is it a problem

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

wojtuss commented:
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/test/pmempool_transform/TEST7, line 56 at r1](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTK7rs1UWHPNy2Pnk3A:-KTnB9El-_imPj2EQTtu:bhm4fne) ([raw file](https://github.com/pmem/nvml/blob/e266c9e1f718b03047c6ee24e5dab384cb14104e/src/test/pmempool_transform/TEST7#L56)):*
<details><summary><i>Previously, lplewa (Łukasz Plewa) wrote…</i></summary>
> Add next functionality to create pool set function to don't modify PATH if is it a problem
</details>
Done.
I've added a requirement for the $DIR path to be an absolute path and removed resolving symlinks in $DIR explicitly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

wojtuss commented:
@plebioda, @lplewa, please take a look at this patch

plebioda commented:
Reviewed 2 of 2 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[src/libpmempool/transform.c, line 104 at r3](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTp42C0dtK8HpvW6zHG:-KTp42C0dtK8HpvW6zHH:b-xvf39w) ([raw file](https://github.com/pmem/nvml/blob/eae6b27c564f9170c0838ccfa3e44d902450aacf/src/libpmempool/transform.c#L104)):*
> ```C
> 			free(path);
> 
> 			/* check if the path does not reoccur in the poolset */
> ```

a separate function for this ? to many nestings...

---

*[src/libpmempool/transform.c, line 110 at r3](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTp49Th39MDG7na-Ma-:-KTp49Th39MDG7na-Ma0:b-v1w8ry) ([raw file](https://github.com/pmem/nvml/blob/eae6b27c564f9170c0838ccfa3e44d902450aacf/src/libpmempool/transform.c#L110)):*
> ```C
> 				unsigned i = (r1 == r2) ? p1 + 1 : 0;
> 				for (unsigned p2 = i; p2 < rep2->nparts; ++p2) {
> 					if (strcmp(PART(rep1, p1).path,
> ```

is that enough ? what if there are different paths to the same file ?

---

*[src/test/unittest/unittest.sh, line 1169 at r3](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTp54FAHMlQRhpXdfe3:-KTp54FBxxw7SkrzyMpE:bhh406w) ([raw file](https://github.com/pmem/nvml/blob/eae6b27c564f9170c0838ccfa3e44d902450aacf/src/test/unittest/unittest.sh#L1169)):*
> ```Shell
> #                          an absolute path
> #
> function require_absolute_path() {
> ```

the require_* functions are for test scripts and normally it means that if specified condition is not met the test should be skipped - not failed. I suggest changing the name.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[src/libpmempool/transform.c, line 110 at r3](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTp49Th39MDG7na-Ma-:-KTuWiz9VlW-oXmUmjdp:bdaek36) ([raw file](https://github.com/pmem/nvml/blob/eae6b27c564f9170c0838ccfa3e44d902450aacf/src/libpmempool/transform.c#L110)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> is that enough ? what if there are different paths to the same file ?
</details>
good point

---

*[src/test/pmempool_transform/TEST7, line 56 at r1](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTK7rs1UWHPNy2Pnk3A:-KTuXgRWGKka1TloRYzv:b-t9gcwk) ([raw file](https://github.com/pmem/nvml/blob/e266c9e1f718b03047c6ee24e5dab384cb14104e/src/test/pmempool_transform/TEST7#L56)):*
<details><summary><i>Previously, wojtuss (Wojciech Uss) wrote…</i></summary>
> Done.
> I've added a requirement for the $DIR path to be an absolute path and removed resolving symlinks in $DIR explicitly.
</details>
I don't think I understand what is the problem here. And I really don't like the require_absolute_path() function. Isn't there any other way to make it work?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

wojtuss commented:
Review status: 2 of 7 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[src/libpmempool/transform.c, line 104 at r3](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTp42C0dtK8HpvW6zHG:-KTrmY4F9LwjyRu2uZyO:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/eae6b27c564f9170c0838ccfa3e44d902450aacf/src/libpmempool/transform.c#L104)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> a separate function for this ? to many nestings...
</details>
Done.

---

*[src/libpmempool/transform.c, line 110 at r3](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTp49Th39MDG7na-Ma-:-KU1C1yygwWNXsneitot:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/eae6b27c564f9170c0838ccfa3e44d902450aacf/src/libpmempool/transform.c#L110)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> good point
</details>
Done.

---

*[src/test/pmempool_transform/TEST7, line 56 at r1](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTK7rs1UWHPNy2Pnk3A:-KU1Pe6OMQVUTK-EKoMp:b-7k2sv2) ([raw file](https://github.com/pmem/nvml/blob/e266c9e1f718b03047c6ee24e5dab384cb14104e/src/test/pmempool_transform/TEST7#L56)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> I don't think I understand what is the problem here. And I really don't like the require_absolute_path() function. Isn't there any other way to make it work?
</details>
create_poolset is already ported to windows, thus a suggestion to use it.
The problem is that test configs on some test machines use symlinks for the $DIR path and create_poolset was resolving them whereas other uses of $DIR in the test was not - comparing the outputs caused the test to fail. So, this change drops resolving symlinks, but preserves the requirement of absolute path for $DIR variable.

---

*[src/test/unittest/unittest.sh, line 1169 at r3](https://reviewable.io:443/reviews/pmem/nvml/1265#-KTp54FAHMlQRhpXdfe3:-KTsfUeSCF3OQmOP5NMI:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/eae6b27c564f9170c0838ccfa3e44d902450aacf/src/test/unittest/unittest.sh#L1169)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> the require_* functions are for test scripts and normally it means that if specified condition is not met the test should be skipped - not failed. I suggest changing the name.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 1 of 2 files at r3, 4 of 5 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[src/common/util_linux.c, line 50 at r4](https://reviewable.io:443/reviews/pmem/nvml/1265#-KU2m5uyKRGNAB4PJ5K8:-KU2m5uyKRGNAB4PJ5K9:b8uq8xo) ([raw file](https://github.com/pmem/nvml/blob/4fa2cb303e06430a73bd964e6b8313c9ab36a81f/src/common/util_linux.c#L50)):*
> ```C
> 
> /*
>  * util_get_full_path -- get full form of the path
> ```

I think, "util_realpath" would be better name

"-- get canonicalized absolute pathname"

---

*[src/common/util_windows.c, line 79 at r4](https://reviewable.io:443/reviews/pmem/nvml/1265#-KU2mqa9dYs8Guhg8SGc:-KU2mqa9dYs8Guhg8SGd:b-hba3fn) ([raw file](https://github.com/pmem/nvml/blob/4fa2cb303e06430a73bd964e6b8313c9ab36a81f/src/common/util_windows.c#L79)):*
> ```C
>  */
> char *
> util_get_full_path(const char *path)
> ```

This function should take a pointer to "resolved_path" buffer.
The caller is responsible to allocate a buffer with a length of at least MAX_PATH. Then, this function would be much simpler.

---

*[src/common/util_windows.c, line 94 at r4](https://reviewable.io:443/reviews/pmem/nvml/1265#-KU2oac5QLV5SVd8U3Wf:-KU2oac5QLV5SVd8U3Wg:b9agl0c) ([raw file](https://github.com/pmem/nvml/blob/4fa2cb303e06430a73bd964e6b8313c9ab36a81f/src/common/util_windows.c#L94)):*
> ```C
> 	} else {
> 		/*
> 		 * the buffer was too small, try again with a buffer large
> ```

Not sure if this would work, but what about passing 0 as the BUFSIZE for the first call to GetFullPathName (with dummy/NULL "buf"), just to obtain the required buffer length? Then malloc() the buffer and call GetFullPathName again - no need for strdup() and on-stack buffers.

---

*[src/common/util_windows.c, line 38 at r5](https://reviewable.io:443/reviews/pmem/nvml/1265#-KUH9Ma8SOWvQsZje4bo:-KUH9Ma8SOWvQsZje4bp:boxq6jz) ([raw file](https://github.com/pmem/nvml/blob/50336950cbe915376f56ec3bf08b1e90c149b92a/src/common/util_windows.c#L38)):*
> ```C
> 
> #include <string.h>
> #include <Windows.h>
> ```

Is this required?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

wojtuss commented:
Review status: 7 of 9 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[src/common/util_linux.c, line 50 at r4](https://reviewable.io:443/reviews/pmem/nvml/1265#-KU2m5uyKRGNAB4PJ5K8:-KUHItHvJO02Nhy_UqCh:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/4fa2cb303e06430a73bd964e6b8313c9ab36a81f/src/common/util_linux.c#L50)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> I think, "util_realpath" would be better name
> 
> "-- get canonicalized absolute pathname"
</details>
Done.

---

*[src/common/util_windows.c, line 79 at r4](https://reviewable.io:443/reviews/pmem/nvml/1265#-KU2mqa9dYs8Guhg8SGc:-KUHdRz4PVeertV4fdOg:b-u1gzkx) ([raw file](https://github.com/pmem/nvml/blob/4fa2cb303e06430a73bd964e6b8313c9ab36a81f/src/common/util_windows.c#L79)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> This function should take a pointer to "resolved_path" buffer.
> The caller is responsible to allocate a buffer with a length of at least MAX_PATH. Then, this function would be much simpler.
> 
</details>
I preserved the api, but simplified the implementation.

---

*[src/common/util_windows.c, line 94 at r4](https://reviewable.io:443/reviews/pmem/nvml/1265#-KU2oac5QLV5SVd8U3Wf:-KUHeIyo3zczYfFlgPfj:bi0qfl) ([raw file](https://github.com/pmem/nvml/blob/4fa2cb303e06430a73bd964e6b8313c9ab36a81f/src/common/util_windows.c#L94)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Not sure if this would work, but what about passing 0 as the BUFSIZE for the first call to GetFullPathName (with dummy/NULL "buf"), just to obtain the required buffer length? Then malloc() the buffer and call GetFullPathName again - no need for strdup() and on-stack buffers.
</details>
Now we are passing BUFSIZE=1, which is sufficient to store only the empty path. That simplifies the function.

---

*[src/common/util_windows.c, line 38 at r5](https://reviewable.io:443/reviews/pmem/nvml/1265#-KUH9Ma8SOWvQsZje4bo:-KUHIuZRnt8EMmxfKZ-e:b32wsm) ([raw file](https://github.com/pmem/nvml/blob/50336950cbe915376f56ec3bf08b1e90c149b92a/src/common/util_windows.c#L38)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Is this required?
</details>
no

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

krzycz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 6 of 6 files at r6.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265#-:-KUI1l1Ch6Dg9FD2JQLQ:bnfp4nl)*
<!-- Sent from Reviewable.io -->

lplewa commented:
Review status: 4 of 9 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/common/util_windows.c, line 78 at r8](https://reviewable.io:443/reviews/pmem/nvml/1265#-KUW7eAes8S6hAopQ1bP:-KUW7eAes8S6hAopQ1bQ:bgic0ea) ([raw file](https://github.com/pmem/nvml/blob/0aedc48a3882c3438bdcd3a6a62ad12aed0918ed/src/common/util_windows.c#L78)):*
> ```C
>  */
> char *
> util_realpath(const char *path)
> ```

I tested this code and it is not working, Also your patch do not include any windows tests...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

wojtuss commented:
Review status: 0 of 16 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[src/common/util_windows.c, line 78 at r8](https://reviewable.io:443/reviews/pmem/nvml/1265#-KUW7eAes8S6hAopQ1bP:-KUuR7oo_ejhrfD2BKBa:b-w94r1q) ([raw file](https://github.com/pmem/nvml/blob/0aedc48a3882c3438bdcd3a6a62ad12aed0918ed/src/common/util_windows.c#L78)):*
<details><summary><i>Previously, lplewa (Łukasz Plewa) wrote…</i></summary>
> I tested this code and it is not working, Also your patch do not include any windows tests...
</details>
Done.
It's working now and a windows test covering the issue is also added.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

plebioda commented:
Please rebase your branch, a fix for failing travis builds has been merged in #1342.

marcinslusarz commented:
Review status: 0 of 19 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/libpmempool/transform.c, line 103 at r10](https://reviewable.io:443/reviews/pmem/nvml/1265#-KV9sKYYLkGV8uoXXaAK:-KV9sKYYLkGV8uoXXaAL:bi1qgst) ([raw file](https://github.com/pmem/nvml/blob/ec49a44c94da397d4723e59d9912a1dc2fd7354b/src/libpmempool/transform.c#L103)):*
> ```C
> 
> /*
>  * check_if_part_path_reoccur -- (internal) check if the part path reoccur later
> ```

s/_path// because this function does not check paths only

Maybe rename it to "is_part_used_once" or something like that?

---

*[src/libpmempool/transform.c, line 148 at r10](https://reviewable.io:443/reviews/pmem/nvml/1265#-KV9tmThRLHS9KfxEuQJ:-KV9tmThRLHS9KfxEuQK:bsogkxn) ([raw file](https://github.com/pmem/nvml/blob/ec49a44c94da397d4723e59d9912a1dc2fd7354b/src/libpmempool/transform.c#L148)):*
> ```C
> /*
>  * check_paths - (internal) check if directories for part files exist
>  *                     and if paths for part files do not repeat in the poolset
> ```

check_parts?

---

*[src/test/pmem_provider/pmem_provider, line 0 at r10](https://reviewable.io:443/reviews/pmem/nvml/1265#-KV9uIE2db37bawvKuwy:-KV9uIE2db37bawvKuwz:b3kre) ([raw file](https://github.com/pmem/nvml/blob/ec49a44c94da397d4723e59d9912a1dc2fd7354b/src/test/pmem_provider/pmem_provider#L0)):*
?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

wojtuss commented:
Review status: 0 of 18 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/libpmempool/transform.c, line 103 at r10](https://reviewable.io:443/reviews/pmem/nvml/1265#-KV9sKYYLkGV8uoXXaAK:-KVOg6OZs46kUzSGsMQv:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/ec49a44c94da397d4723e59d9912a1dc2fd7354b/src/libpmempool/transform.c#L103)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> s/_path// because this function does not check paths only
> 
> Maybe rename it to "is_part_used_once" or something like that?
> 
</details>
Done.

---

*[src/libpmempool/transform.c, line 148 at r10](https://reviewable.io:443/reviews/pmem/nvml/1265#-KV9tmThRLHS9KfxEuQJ:-KVA60JQQhHULNK2tvlo:bs3fe8w) ([raw file](https://github.com/pmem/nvml/blob/ec49a44c94da397d4723e59d9912a1dc2fd7354b/src/libpmempool/transform.c#L148)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> check_parts?
</details>
no, we do check paths, not parts themselves

---

*[src/test/pmem_provider/pmem_provider, line  at r10](https://reviewable.io:443/reviews/pmem/nvml/1265#-KV9uIE2db37bawvKuwy:-KVOgHMQWGvj-3gmQDxt:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/ec49a44c94da397d4723e59d9912a1dc2fd7354b/src/test/pmem_provider/pmem_provider#L)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> ?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

plebioda commented:
*[doc/libpmempool.3.md, line 353 at r11](https://reviewable.io:443/reviews/pmem/nvml/1265#-KV_LoqRuDB9iRkMohsG:-KV_LoqRuDB9iRkMohsH:bzaminw) ([raw file](https://github.com/pmem/nvml/blob/d8893c0835a942f65e0505085ba885a11188c1d5/doc/libpmempool.3.md#L353)):*
> ```Markdown
> 
> When adding or deleting replica, the two poolset files can differ only in the
> definitions of replicas which are to be added or deleted. Only one of the
> ```

I think you should state it clearly that you can add or remove multiple replicas in the same operation. Also please make a note about support for reordering.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

plebioda commented:
*[src/common/util_linux.c, line 69 at r11](https://reviewable.io:443/reviews/pmem/nvml/1265#-KV_MoERm_prud63zWgN:-KV_MoERm_prud63zWgO:b-98aior) ([raw file](https://github.com/pmem/nvml/blob/d8893c0835a942f65e0505085ba885a11188c1d5/src/common/util_linux.c#L69)):*
> ```C
> 	struct stat sb1, sb2;
> 	if (stat(path1, &sb1) || stat(path2, &sb2)) {
> 		LOG(2, "stat failed for %s or %s", path1, path2);
> ```

come on :)

if (stat(path1, &sb1)) {
    LOG(2, "!stat for $s failed, using strcmp", path1);
    return strcmp(path1, path2);
}

if (stat(path2, &sb2)) {
    LOG(2, "!stat for $s failed, using strcmp", path2);
    return strcmp(path1, path2);
}

return sb1.st_dev != sb2.st_dev || sb1.st_ino != sb2.st_ino;

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

plebioda commented:
Reviewed 6 of 6 files at r6, 2 of 2 files at r7, 15 of 17 files at r9, 2 of 2 files at r11.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[src/common/util_linux.c, line 69 at r11](https://reviewable.io:443/reviews/pmem/nvml/1265#-KV_MoERm_prud63zWgN:-KV_NmJvwgEkHihm25Uf:b-fkj4ok) ([raw file](https://github.com/pmem/nvml/blob/d8893c0835a942f65e0505085ba885a11188c1d5/src/common/util_linux.c#L69)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> come on :)
> 
> if (stat(path1, &sb1)) {
>     LOG(2, "!stat for $s failed, using strcmp", path1);
>     return strcmp(path1, path2);
> }
> 
> if (stat(path2, &sb2)) {
>     LOG(2, "!stat for $s failed, using strcmp", path2);
>     return strcmp(path1, path2);
> }
> 
> return sb1.st_dev != sb2.st_dev || sb1.st_ino != sb2.st_ino;
> 
> 
</details>
Plus I think we should only fallback to strcmp if the stat returns ENOENT ?

---

*[src/common/util_windows.c, line 88 at r7](https://reviewable.io:443/reviews/pmem/nvml/1265#-KUMtwoU4JkI_PHI5E_8:-KUMtwoU4JkI_PHI5E_9:b-ofo3os) ([raw file](https://github.com/pmem/nvml/blob/1077535786a4ad10cc53e5fc505ea26e36449c18/src/common/util_windows.c#L88)):*
> ```C
> 		/* the function failed */
> 		return NULL;
> 	} else {
> ```

else and intentation not needed

---

*[src/libpmempool/transform.c, line 103 at r7](https://reviewable.io:443/reviews/pmem/nvml/1265#-KUMuAtDrIsN1QTUMa-7:-KUMuAtDrIsN1QTUMa-8:b-dophl8) ([raw file](https://github.com/pmem/nvml/blob/1077535786a4ad10cc53e5fc505ea26e36449c18/src/libpmempool/transform.c#L103)):*
> ```C
> 
> /*
>  * check_if_part_path_reoccur -- (internal) check if the part path reoccur later
> ```

check_path_duplicate ? 

---

*[src/libpmempool/transform.c, line 124 at r7](https://reviewable.io:443/reviews/pmem/nvml/1265#-KUMuN9eoCt47rooCflT:-KUMuN9eoCt47rooCflU:b7p6wq9) ([raw file](https://github.com/pmem/nvml/blob/1077535786a4ad10cc53e5fc505ea26e36449c18/src/libpmempool/transform.c#L124)):*
> ```C
> 			char *pathp = util_realpath(PART(repr, p).path);
> 			if (pathp == NULL) {
> 				LOG(1, "Cannot get absolute path for %s,"
> ```

is that possible ? I mean I think we should fail in this case.

---

*[src/libpmempool/transform.c, line 112 at r11](https://reviewable.io:443/reviews/pmem/nvml/1265#-KV_OBxQwEz-hazbt0f7:-KV_OBxQwEz-hazbt0f8:b-opa4m1) ([raw file](https://github.com/pmem/nvml/blob/d8893c0835a942f65e0505085ba885a11188c1d5/src/libpmempool/transform.c#L112)):*
> ```C
> 	char *path = util_realpath(PART(rep, partn).path);
> 	if (path == NULL) {
> 		LOG(1, "Cannot get absolute path for %s, replica %u, part %u",
> ```

please use lowercase

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 2 of 2 files at r7, 16 of 17 files at r9, 2 of 2 files at r11.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

wojtuss commented:
Review status: 12 of 18 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[doc/libpmempool.3.md, line 353 at r11](https://reviewable.io:443/reviews/pmem/nvml/1265#-KV_LoqRuDB9iRkMohsG:-KVeHluHrJi22A4TXJDO:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/d8893c0835a942f65e0505085ba885a11188c1d5/doc/libpmempool.3.md#L353)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> I think you should state it clearly that you can add or remove multiple replicas in the same operation. Also please make a note about support for reordering.
</details>
Done.

---

*[src/common/util_linux.c, line 69 at r11](https://reviewable.io:443/reviews/pmem/nvml/1265#-KV_MoERm_prud63zWgN:-KVeQy3Eo7z1TQS0UqQD:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/d8893c0835a942f65e0505085ba885a11188c1d5/src/common/util_linux.c#L69)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> Plus I think we should only fallback to strcmp if the stat returns ENOENT ?
</details>
Done.

---

*[src/common/util_windows.c, line 88 at r7](https://reviewable.io:443/reviews/pmem/nvml/1265#-KUMtwoU4JkI_PHI5E_8:-KVeI3vcVdDyP7sC7tlf:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/1077535786a4ad10cc53e5fc505ea26e36449c18/src/common/util_windows.c#L88)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> else and intentation not needed
</details>
Done.

---

*[src/libpmempool/transform.c, line 103 at r7](https://reviewable.io:443/reviews/pmem/nvml/1265#-KUMuAtDrIsN1QTUMa-7:-KVeJiW0yJW1qvhyE12B:b-74k4be) ([raw file](https://github.com/pmem/nvml/blob/1077535786a4ad10cc53e5fc505ea26e36449c18/src/libpmempool/transform.c#L103)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> check_path_duplicate ? 
</details>
nah :-)

---

*[src/libpmempool/transform.c, line 124 at r7](https://reviewable.io:443/reviews/pmem/nvml/1265#-KUMuN9eoCt47rooCflT:-KVeOmkuYegJ5ZBzBU_-:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/1077535786a4ad10cc53e5fc505ea26e36449c18/src/libpmempool/transform.c#L124)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> is that possible ? I mean I think we should fail in this case.
</details>
Done.

---

*[src/libpmempool/transform.c, line 112 at r11](https://reviewable.io:443/reviews/pmem/nvml/1265#-KV_OBxQwEz-hazbt0f7:-KVeK_9KijRGUO-jT0wN:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/d8893c0835a942f65e0505085ba885a11188c1d5/src/libpmempool/transform.c#L112)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> please use lowercase
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

plebioda commented:
Reviewed 2 of 8 files at r12.
Review status: 12 of 18 files reviewed at latest revision, 13 unresolved discussions.

---

*[src/common/util_linux.c, line 80 at r12](https://reviewable.io:443/reviews/pmem/nvml/1265#-KW9lTQ4OtUOUVZcpEuR:-KW9lTQ4OtUOUVZcpEuS:brzva21) ([raw file](https://github.com/pmem/nvml/blob/833c8ba477a710dd1d2b9c127dd673379b2fffe3/src/common/util_linux.c#L80)):*
> ```C
> 	if (stat(path2, &sb2)) {
> 		if (errno != ENOENT) {
> 			ERR("stat failed for %s, errno %d", path2, errno);
> ```

if you start error message with '!' (exclamation mark) the error message will be printed using strerror - there is no point in printing errno number.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

wojtuss commented:
Review status: 11 of 18 files reviewed at latest revision, 13 unresolved discussions.

---

*[src/common/util_linux.c, line 80 at r12](https://reviewable.io:443/reviews/pmem/nvml/1265#-KW9lTQ4OtUOUVZcpEuR:-KWCxeJDeEvcgRvry-SW:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/833c8ba477a710dd1d2b9c127dd673379b2fffe3/src/common/util_linux.c#L80)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> if you start error message with '!' (exclamation mark) the error message will be printed using strerror - there is no point in printing errno number.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

plebioda commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 6 of 8 files at r12, 1 of 1 files at r13.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[doc/pmempool-transform.1.md, line 62 at r13](https://reviewable.io:443/reviews/pmem/nvml/1265#-KWY95hRLnQVuPjI9sXO:-KWY95hRLnQVuPjI9sXP:b2ni9fl) ([raw file](https://github.com/pmem/nvml/blob/2cca36efbc3517dc667bab15700709a59e16706b/doc/pmempool-transform.1.md#L62)):*
> ```Markdown
> The following operations are supported:
> 
> * adding replicas - on or more new replicas can be added and synchronized with
> ```

one

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265#-:-KWYAJE8v8I8rHul-WfD:bnfp4nl)*
<!-- Sent from Reviewable.io -->

plebioda commented:
Reviewed 16 of 17 files at r14.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

wojtuss commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[doc/pmempool-transform.1.md, line 62 at r13](https://reviewable.io:443/reviews/pmem/nvml/1265#-KWY95hRLnQVuPjI9sXO:-KWcDh2dLxZJ_e-948hB:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/2cca36efbc3517dc667bab15700709a59e16706b/doc/pmempool-transform.1.md#L62)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> one
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

plebioda commented:
Reviewed 1 of 17 files at r14, 3 of 4 files at r15, 2 of 2 files at r16, 3 of 3 files at r17, 7 of 7 files at r18.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[src/common/file_windows.c, line 107 at r18](https://reviewable.io:443/reviews/pmem/nvml/1265#-KWlbwUiJXDGkcZCllHW:-KWlbwUiJXDGkcZCllHX:b2nh4z9) ([raw file](https://github.com/pmem/nvml/blob/cfa0743390b52f1da82cbe3822e5152be7d83364/src/common/file_windows.c#L107)):*
> ```C
> 
> /*
>  * util_is_absolute_path -- check if the path is absolute;
> ```

; ?

---

*[src/libpmempool/replica.c, line 955 at r18](https://reviewable.io:443/reviews/pmem/nvml/1265#-KWlcRUMDSCn5SwX3GWC:-KWlcRUMDSCn5SwX3GWD:blu5i55) ([raw file](https://github.com/pmem/nvml/blob/cfa0743390b52f1da82cbe3822e5152be7d83364/src/libpmempool/replica.c#L955)):*
> ```C
> 
> /*
>  * check_if_part_dir_exists -- (internal) check if directory for the part file
> ```

please update comment

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

lplewa commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265#-:-KWlehxDeBuiTunQmmxM:bnfp4nl)*
<!-- Sent from Reviewable.io -->

wojtuss commented:
Review status: 25 of 28 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/common/file_windows.c, line 107 at r18](https://reviewable.io:443/reviews/pmem/nvml/1265#-KWlbwUiJXDGkcZCllHW:-KWlfOaSHC3YD4ageT9V:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/cfa0743390b52f1da82cbe3822e5152be7d83364/src/common/file_windows.c#L107)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> ; ?
</details>
Done.

---

*[src/libpmempool/replica.c, line 955 at r18](https://reviewable.io:443/reviews/pmem/nvml/1265#-KWlcRUMDSCn5SwX3GWC:-KWlf5QOE4c3GHlaHq4G:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/cfa0743390b52f1da82cbe3822e5152be7d83364/src/libpmempool/replica.c#L955)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> please update comment
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 1 of 8 files at r12, 1 of 1 files at r13, 11 of 17 files at r14, 3 of 4 files at r15, 3 of 3 files at r17, 6 of 7 files at r18, 3 of 3 files at r19.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[src/libpmempool/replica.c, line 965 at r19](https://reviewable.io:443/reviews/pmem/nvml/1265#-KWm6Ub6Rk6LcIVqI5ML:-KWm6Ub6Rk6LcIVqI5MM:b9bkckz) ([raw file](https://github.com/pmem/nvml/blob/30289a71122e9e8db79d9ab1d3c7653df967e22f/src/libpmempool/replica.c#L965)):*
> ```C
> 	char *path = Strdup(PART(REP(set, repn), partn).path);
> 	const char *dir = dirname(path);
> 	struct stat sb;
> ```

util_stat_t / util_stat()
(was already fixed, now reverted)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

plebioda commented:
Reviewed 1 of 17 files at r14, 3 of 3 files at r19.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[src/libpmempool/replica.c, line 955 at r18](https://reviewable.io:443/reviews/pmem/nvml/1265#-KWlcRUMDSCn5SwX3GWC:-KWm8SMNrIJlht6xP66J:b67ppjv) ([raw file](https://github.com/pmem/nvml/blob/cfa0743390b52f1da82cbe3822e5152be7d83364/src/libpmempool/replica.c#L955)):*
<details><summary><i>Previously, wojtuss (Wojciech Uss) wrote…</i></summary>
> Done.
</details>
not done. the function name does not match.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

wojtuss commented:
Review status: 27 of 28 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/libpmempool/replica.c, line 955 at r18](https://reviewable.io:443/reviews/pmem/nvml/1265#-KWlcRUMDSCn5SwX3GWC:-KWmUOFdwpGk5dVbsDXu:b-4ohf7u) ([raw file](https://github.com/pmem/nvml/blob/cfa0743390b52f1da82cbe3822e5152be7d83364/src/libpmempool/replica.c#L955)):*
<details><summary><i>Previously, plebioda wrote…</i></summary>
> not done. the function name does not match.
</details>
sorry, now done.

---

*[src/libpmempool/replica.c, line 965 at r19](https://reviewable.io:443/reviews/pmem/nvml/1265#-KWm6Ub6Rk6LcIVqI5ML:-KWmV0ZY0AKs7v4t2fMq:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/30289a71122e9e8db79d9ab1d3c7653df967e22f/src/libpmempool/replica.c#L965)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> util_stat_t / util_stat()
> (was already fixed, now reverted)
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265)*
<!-- Sent from Reviewable.io -->

plebioda commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r20.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265#-:-KWmYEP4ITJ6EnCw6diS:bnfp4nl)*
<!-- Sent from Reviewable.io -->

krzycz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r20.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1265#-:-KWm_anYExF-bLE2OBkS:bnfp4nl)*
<!-- Sent from Reviewable.io -->
